src/abstractdataframe/iteration.jl

```diff
@@ -72,29 +75,73 @@ julia> df = DataFrame(x=1:4, y=11:14)
 │ 3   │ 3     │ 13    │
 │ 4   │ 4     │ 14    │
 
-julia> map(sum, eachcol(df))
-1×2 DataFrame
+julia> collect(eachcol(df))
+2-element Array{Tuple{Symbol,Any},1}:
+ (:x, [1, 2, 3, 4])
+ (:y, [11, 12, 13, 14])
+```
+"""
+eachcol(df::T) where T<: AbstractDataFrame = DFColumnIterator{T, true}(df)
+
+"""
+    columns(df::AbstractDataFrame)
+
+Return a `DFColumnIterator` that iterates an `AbstractDataFrame` column by column.
+Iteration returns a column vector.
+
+**Examples**
+
+```jldoctest
+julia> df = DataFrame(x=1:4, y=11:14)
+4×2 DataFrame
 │ Row │ x     │ y     │
 │     │ Int64 │ Int64 │
 ├─────┼───────┼───────┤
-│ 1   │ 10    │ 50    │
+│ 1   │ 1     │ 11    │
+│ 2   │ 2     │ 12    │
+│ 3   │ 3     │ 13    │
+│ 4   │ 4     │ 14    │
+
+julia> collect(columns(df))
+2-element Array{AbstractArray{T,1} where T,1}:
+ [1, 2, 3, 4]
+ [11, 12, 13, 14]
 ```
 """
-eachcol(df::AbstractDataFrame) = DFColumnIterator(df)
+columns(df::T) where T<: AbstractDataFrame = DFColumnIterator{T, false}(df)
 
-function Base.iterate(itr::DFColumnIterator, j=1)
+Base.length(itr::DFColumnIterator) = size(itr.df, 2)
+Base.size(itr::DFColumnIterator) = (size(itr.df, 2),)
+Base.size(itr::DFColumnIterator, ix) =
+    ix == 1 ? length(itr) : throw(ArgumentError("Incorrect dimension"))
+Base.firstindex(itr::DFColumnIterator) = 1
+Base.lastindex(itr::DFColumnIterator) = length(itr)
+
+function Base.iterate(itr::DFColumnIterator{<:AbstractDataFrame,true}, j=1)
     j > size(itr.df, 2) && return nothing
     return ((_names(itr.df)[j], itr.df[j]), j + 1)
 end
-Base.eltype(::DFColumnIterator) = Tuple{Symbol, AbstractVector}
-Base.size(itr::DFColumnIterator) = (size(itr.df, 2), )
-Base.length(itr::DFColumnIterator) = size(itr.df, 2)
-Base.getindex(itr::DFColumnIterator, j) = itr.df[j]
-function Base.map(f::Union{Function,Type}, dfci::DFColumnIterator)
-    # note: `f` must return a consistent length
-    res = DataFrame()
-    for (n, v) in eachcol(dfci.df)
-        res[n] = f(v)
+
+Base.eltype(::DFColumnIterator{<:AbstractDataFrame,true}) =
+    Tuple{Symbol, AbstractVector}
+
+function Base.getindex(itr::DFColumnIterator{<:AbstractDataFrame,true}, j)
+    # TODO: change to the way getindex for false is defined below afted deprecation
+    if !(j isa Integer) || j isa Bool
+        Base.depwarn("calling getindex on DFColumnIterator{<:AbstractDataFrame,true} " *
+                     " object will only accept integer indexing and will return " *
```

Comment:
Would be nice to advise using `columns` instead of `eachcol`. `DFColumnIterator` is an implementation detail which doesn't need to be mentioned. Also, remove leading space.

Reply:
fixed